### DISCUSSION TITLE
Private Bitbucket packages try to fetch from packagist

### DISCUSSION THREAD

**Discussion by BernhardK91** [renovatebot/renovate#24065] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.42.4

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Cloud (bitbucket.org)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

We have some private hosted Composer (PHP) packages, hosted on Bitbucket. Unfortunately, renovate thinks they are packagist repos and tries to fetch from there, what fails.

I want to have renovate directly to check Bitbucket and not try packagist instead.

How can I tell renovate to use "bitbucket-tags" instead?

### What you tried so far.

Tried to play with packageRules and adding a name to the "repositories" in the `composer.json`. Nothing worked so far.

### Relevant debug logs

```
GET https://packagist.org/p2/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE]~dev.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=49) (repository=XXXXXX, baseBranch=development)
```

```
{
                 "depType": "require",
                 "depName": "[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE]",
                 "currentValue": "*",
                 "lockedVersion": "0.2.1",
                 "datasource": "packagist",
                 "registryUrls": [
                   "https://repo.backpackforlaravel.com/",
                   "https://packagist.org/"
                 ],
                 "updates": [],
                 "packageName": "[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE]",
                 "versioning": "composer",
                 "warnings": [
                   {
                     "topic": "[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE]",
                     "message": "Failed to look up packagist package [MY-WORKSPACE]/[MY-PRIVATE-PACKAGE]"
                   }
                 ]
               },
```

**Comment by rarkins** [comment]
Is `https://repo.backpackforlaravel.com/` correct or is that also incorrect?

**Reply by BernhardK91** [reply]
The backpack URL is correct. Here is the definition from the composer.json:

```
"repositories": [
        {
            "type": "composer",
            "url": "https://repo.backpackforlaravel.com/"
        },
        {
            "type": "vcs",
            "name": "[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE]",
            "url": "https://bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git"
        },
    ],
```

By adding the name attribute (found in one of the issues and in the [docs](https://docs.renovatebot.com/modules/manager/composer/#additional-information)) we figured out, that it uses `git-tags` now.

Currently, we are "only" struggling with the authentication on Bitbucket:

```
DEBUG: Datasource unknown error (repository=XXXXXX, baseBranch=development)
       "datasource": "git-tags",
       "packageName": "https://bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git",
       "err": {
         "task": {
           "commands": [
             "ls-remote",
             "https://**redacted**@bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git"
           ],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "remote: Invalid credentials\nfatal: Authentication failed for 'https://bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git/'\n",
         "stack": "Error: remote: Invalid credentials\nfatal: Authentication failed for 'https://bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git/'\n\n    at Object.action (/opt/containerbase/tools/renovate/36.42.4/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/opt/containerbase/tools/renovate/36.42.4/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /opt/containerbase/tools/renovate/36.42.4/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/opt/containerbase/tools/renovate/36.42.4/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/opt/containerbase/tools/renovate/36.42.4/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/opt/containerbase/tools/renovate/36.42.4/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
       }
DEBUG: Failed to look up git-tags package https://bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git (repository=XXXXX, baseBranch=development, packageFile=composer.json, dependency=https://bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git)
```

We tried by adding credentials (tried username and app-password and Auth Token) in to the config.js of the "renovate server config":

```
hostRules: [
        {
            matchHost: "https://bitbucket.org/[MY-WORKSPACE]/",
            token: process.env.BITBUCKET_BASIC_AUTH_TOKEN,
            authType: 'Basic',
        }
    ],
```

Any suggestions, what I missed?

Thank you very much!

**Reply by rarkins** [reply]
Are you able to successfully run `git ls-remote https://SOMECREDENTIALS@bitbucket.org/[MY-WORKSPACE]/[MY-PRIVATE-PACKAGE].git` from your own PC? I don't know if it needs to be a token or a username:password

**Reply by BernhardK91** [reply]
Yes. With username:app-password from local it works fine:

![image](https://github.com/renovatebot/renovate/assets/5526334/12791bab-71e8-458a-b7b7-9fad97dec5ff)

**Reply by rarkins** [reply]
The git-tags datasource is here: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/git-tags/index.ts

It delegates to the git-refs datasource: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/git-refs/index.ts

The only auth for this datasource is here: https://github.com/renovatebot/renovate/blob/151435acfd1e96d8591935b437e5604e38f75fbf/lib/modules/datasource/git-refs/base.ts#L28-L33

And it looks like that filters out any hostRules which don't contain a _token_ specifically: https://github.com/renovatebot/renovate/blob/151435acfd1e96d8591935b437e5604e38f75fbf/lib/util/git/auth.ts#L173-L176

There is actually a bitbucket-tags datasource here: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/bitbucket-tags/index.ts but when extracted from Composer, it's not being detected

Unfortunately without a reproduction, and without being a regular Composer or Bitbucket Cloud user, I don't have any way to debug it at runtime to see which paths it's hitting.

If you or anyone else with the same problem in future wants to take a stab at it, adding logic like https://github.com/renovatebot/renovate/blob/151435acfd1e96d8591935b437e5604e38f75fbf/lib/modules/manager/flux/extract.ts#L136-L142 to the composer manager may be enough to (a) switch the dep from git-tags to bitbucket-tags, and then (b) leverage the hostRules approach already built into bitbucket-tags

**Reply by viceice** [reply]
@rarkins Better add that logic to the `git-tags` datasource and forward from there if match? so all managers benefit out of the box?

**Reply by rarkins** [reply]
Yep, see #19459

**Reply by BernhardK91** [reply]
I think the best way would be to make sure when in composer it is not "vcs" but "bitbucket", then it should use the bitbucket-tags. From my current understanding, that would be the best and most sustainable solution.

**Reply by BernhardK91** [reply]
I have tried it locally with the docker image (didn't work) and direct from the sources (worked). 

That made me think, it must work, when I have a valid SSH key with access to the repo available.

EDIT: It had nothing to do with SSH keys. Windows stored the git credentials for Bitbucket in the `Windows Credential Manager`. That just showed again that username and app-password work.

**Reply by BernhardK91** [reply]
I have created a PR https://github.com/renovatebot/renovate/pull/24096 that adds using bitbucket-tags for Bitbucket repos.

**Reply by BernhardK91** [reply, marked as answer]
The solution to the original problem was adding "name" attribute in the repositories config of the `composer.json`. Since version `36.87.0` Bitbucket repositories (also private ones) are working with Renovate as well.
